src/Compilers/Core/Portable/Symbols/ISymbolExtensions_IsAccessibleWithin.cs

```diff
@@ -0,0 +1,291 @@
+                    throw ExceptionUtilities.UnexpectedValue(symbol.Kind);
+            }
+        }
+
+        private static bool IsNamedTypeAccessibleWithin(INamedTypeSymbol type, ISymbol within, ITypeSymbol throughTypeOpt)
+        {
+            Debug.Assert(within is INamedTypeSymbol || within is IAssemblySymbol);
+            Debug.Assert(type != null);
+
+            if (!type.IsDefinition)
+            {
+                foreach (var typeArg in type.TypeArguments)
+                {
+                    if (!IsAccessibleWithin(typeArg, within, throughTypeOpt))
+                    {
+                        return false;
+                    }
+                }
+            }
+
+            var containingType = type.ContainingType;
+            if (containingType == null)
+            {
+                return IsNonNestedTypeAccessible(type.ContainingAssembly, type.DeclaredAccessibility, within);
+            }
+            else
+            {
+                return IsMemberAccessible(containingType, type.DeclaredAccessibility, within, null);
+            }
+        }
+
+        private static bool IsNonNestedTypeAccessible(IAssemblySymbol assembly, Accessibility declaredAccessibility, ISymbol within)
+        {
+            Debug.Assert(within is INamedTypeSymbol || within is IAssemblySymbol);
+            Debug.Assert(assembly != null);
+
+            switch (declaredAccessibility)
+            {
+                case Accessibility.NotApplicable:
+                case Accessibility.Public:
+                    // Public symbols are always accessible from any context
+                    return true;
+
+                case Accessibility.Private:
+                case Accessibility.Protected:
+                case Accessibility.ProtectedAndInternal:
+                    // Shouldn't happen except in error cases as these access levels cannot be used to declare top-level types.
+                    return false;
+
+                case Accessibility.Internal:
+                case Accessibility.ProtectedOrInternal:
+
+                    // An internal type is accessible if we're in the same assembly or we have
+                    // friend access to the assembly it was defined in.
+                    var withinAssembly = (within as IAssemblySymbol) ?? ((INamedTypeSymbol)within).ContainingAssembly;
+                    return withinAssembly.HasInternalAccessTo(assembly);
+
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(declaredAccessibility);
+            }
+        }
+
+        private static bool HasInternalAccessTo(this IAssemblySymbol fromAssembly, IAssemblySymbol toAssembly)
+        {
+            if (fromAssembly == toAssembly)
+            {
+                return true;
+            }
+
+            if (fromAssembly.IsInteractive && toAssembly.IsInteractive)
+            {
+                return true;
+            }
+
+            return toAssembly.GivesAccessTo(fromAssembly);
```

Comment:
>GivesAccessTo [](start = 30, length = 13)

It looks like this is never true when toAssembly and fromAssembly belong to different languages.

Reply:
Sounds like a bug. Either an assembly in one language is permitted to give access to an assembly in another language, in which case the API is returning an incorrect result, or an assembly in one language is *not* permitted to give access to an assembly in another language, in which case this code is correct. Either way, if there is a problem it is in the implementation of `toAssembly.GivesAccessTo`, not here.

Reply:
> in which case this code is correct

Perhaps we should get clarity and agree on what does it mean for this component to behave correctly, what are the expected results in various scenarios. Your response makes me to believe that even you are not sure what to expect.

Reply:
I agree it would be good to get clarity on all the relational cases.  This will help ensure everyone is on the same page and we all think it makes sense and can actually be implemented.  It will also be helpful toward driving a suitable test suite that then validates these expectations.  

From my reading of the code, the areas that need clarity are:

1. Internal, and how accessibilty across different-language-boundary should be defined.
2. protected, and how inheritance across different-language-boundary be defined.

Technically, it looks like 'private' also does type checks.  However, private seems well defined across languages/assemblies  It's always false as something in another assembly (which you would always get if you had a different language) would never be able to see your privates.

Reply:
> Your response makes me to believe that even you are not sure what to expect.

I am not sure whether there is a bug in the language-specific `Symbol` APIs, since you reported them not obeying their specifications. This API assumes that the Symbol APIs are correct.

I expect the `Symbol` APIs to return correct results, in which case this API would return correct results.

Reply:
>  am not sure whether there is a bug in the language-specific Symbol APIs, since you reported them not obeying their specifications.

Whether this is a bug or intentional designed behavior to keep language worlds separate is a separate discussion.

> I expect the Symbol APIs to return correct results, in which case this API would return correct results.

Where is a test for you assumptions and expectations? API do not have to behave the way you assume them to behave.

Reply:
I'll report a bug that the language-specific implementations do not obey their contract.

---
In reply to: [184278437](https://github.com/dotnet/roslyn/pull/26367#discussion_r184278437) [](ancestors = 184278437)

Reply:
> would you be willing to go investigate these and to come back with a proposal that we can discuss

Are you asking me to transcribe the CLR access rules? The only "variable" is how we define type and assembly identity. Assembly identity is supposed to be defined by` IAssemblySymbol.Identity` but @AlekseyTs points out that the compiler does it wrong. Type identity should be defined, recursively, by name and arity, and containing symbol. The rest is down to the CLR rules for the meaning of the access modes.

---
In reply to: [184489272](https://github.com/dotnet/roslyn/pull/26367#discussion_r184489272) [](ancestors = 184489272)

Reply:
@gafter Thanks!

Reply:
We don't have tests for interfaces, because we don't have implementations in the interfaces.

---
In reply to: [184506449](https://github.com/dotnet/roslyn/pull/26367#discussion_r184506449) [](ancestors = 184506449)

Reply:
The code no longer compares assemblies using ==. We compare assembly identities instead, which is the CLR rule.

---
In reply to: [184566431](https://github.com/dotnet/roslyn/pull/26367#discussion_r184566431) [](ancestors = 184566431,184278437)

Reply:
The implementation now compares assemblies by comparing their identities.

---
In reply to: [184567028](https://github.com/dotnet/roslyn/pull/26367#discussion_r184567028) [](ancestors = 184567028,184489272)